Write a compiler from Scrapscript to C

tekknolagi opened the pull request:
Emit straight C, no IR. Includes a GC and small runtime.

Has immediate small integers (no bignum), small strings and heap strings, and immediate empty list.

tekknolagi committed:
Sketch of compiling numbers

tekknolagi committed:
Multiply, print

tekknolagi committed:
Remove return of trace_roots

tekknolagi committed:
Fix syntax error and add more handles

tekknolagi committed:
Add lists

tekknolagi committed:
Print lists and println

tekknolagi committed:
Put expression in list example

tekknolagi committed:
Clean up heap at end

tekknolagi committed:
Simplify to printf

tekknolagi committed:
Support user-defined functions without closure variables

tekknolagi committed:
Assert

tekknolagi committed:
Support list in collect

tekknolagi committed:
Add closures to runtime

tekknolagi committed:
Drop little hints on allocate

tekknolagi committed:
WIP everything is a closure

tekknolagi committed:
Add TODO

tekknolagi committed:
Pass this argument to closures

tekknolagi committed:
Root closure_get temporaries

tekknolagi committed:
Smaller demo with freevars

tekknolagi committed:
Silence warning

tekknolagi committed:
Use X-macro and __COUNTER__ for tag values

tekknolagi committed:
Remove unused cons (for now?)

tekknolagi committed:
Move source of truth for variable-sized objects

tekknolagi committed:
Split up allocator slow path

tekknolagi committed:
Collapse union to avoid UB

tekknolagi committed:
Add some asserts

tekknolagi committed:
Remove printf in forward()

tekknolagi committed:
Add debug memset

tekknolagi committed:
nit

tekknolagi committed:
Run tests only if scrapscript.py modified

tekknolagi committed:
Add debug collects

tekknolagi committed:
nit

tekknolagi committed:
WIP: Preliminary match/case

tekknolagi committed:
Widen num; add getter

tekknolagi committed:
Use num getter in compiler

tekknolagi committed:
Add number subtraction

tekknolagi committed:
Allow named functions to call themselves

tekknolagi committed:
Don't fall-through in Var case; return early

tekknolagi committed:
Add sample match function

tekknolagi committed:
Simplify sample code

tekknolagi committed:
Restore original names

tekknolagi committed:
Add some list runtime functions

tekknolagi committed:
Make sure to GC protect all parameters; they are raw pointers

tekknolagi committed:
Add list_append runtime function

tekknolagi committed:
Add factorial into the mix, why not

tekknolagi committed:
Provide known name for entrypoint

tekknolagi committed:
Add TODO about const heap

tekknolagi committed:
Add input filename and -o and --format

tekknolagi committed:
Add --compile and --run

tekknolagi committed:
Generate builtin wrappers from compiler

tekknolagi committed:
Check format/compile/run output

tekknolagi committed:
Fix closure/runtime contract

tekknolagi committed:
Use CompiledFunction machinery

tekknolagi committed:
Don't cast

tekknolagi committed:
Assert closure before attempting to call

tekknolagi committed:
Use function to retrieve closure fn

tekknolagi committed:
Add TODO

tekknolagi committed:
Remove finished TODO

tekknolagi committed:
Use new OBJECT_HANDLE for maybe more readability

tekknolagi committed:
Split pattern and body

tekknolagi committed:
Allocate and print records

Records are stored in pairs of {key_idx, value_obj} where key_idx is the
index of the globally interned record key name.

tekknolagi committed:
Record access and guards

tekknolagi committed:
Default to printing result of main

tekknolagi committed:
Use is_list and add list_get

tekknolagi committed:
Support record in GC

tekknolagi committed:
Enable GC stress and other checks by default

tekknolagi committed:
Oops, move record below closure

tekknolagi committed:
Add some more guards

tekknolagi committed:
Remove done TODO

tekknolagi committed:
Make compiler.py executable

tekknolagi committed:
WIP: pattern matching redux with chris

tekknolagi committed:
WIP bind variables

tekknolagi committed:
Fix some naming and syntax issues but IT WORKS

tekknolagi committed:
list cons, list_rest, spread match

tekknolagi committed:
Cleanup

tekknolagi committed:
Make heap size configurable

tekknolagi committed:
Just prepend runtime.c instead of including it

tekknolagi committed:
Add handles bound check

tekknolagi committed:
Name user functions in C if known

tekknolagi committed:
Avoid formatting in print when not needed

tekknolagi committed:
Add debug mode to the compiler

Change flags and also choose whether or not to emit extra GC collect()

tekknolagi committed:
Support immediate objects

Make numbers immediate (still need to support bignums later).

Introduce `struct object*` which can either be immediate (low bit is 0)
or heap-allocated (low bit is 1). Only heap-allocated objects can be
cast to `struct gc_obj*`.

Add a lot of helper functions because enforcing these invariants
otherwise is very failure-prone.

tekknolagi committed:
Clean up

tekknolagi committed:
Support compiling with cosmocc

tekknolagi committed:
Support LDFLAGS

tekknolagi committed:
Add docs for compiling

tekknolagi committed:
Refactor flags handling

tekknolagi committed:
nit

tekknolagi committed:
nit

tekknolagi committed:
Fix some lints

tekknolagi committed:
nit

tekknolagi committed:
Add more handles

tekknolagi committed:
Pass record_field directly to record_set

tekknolagi committed:
Make lists cons cells

We pattern match in a way that makes vector-type lists very slow and not
so useful.

tekknolagi committed:
Add heap strings

tekknolagi committed:
Split string into small_string and heap_string

Store small strings (0-7 bytes) in a pointer instead of on the heap.

tekknolagi committed:
Add string concat

tekknolagi committed:
Fiddle with inlining

tekknolagi committed:
Make indices unsigned

tekknolagi committed:
Bring over ALWAYS_INLINE

tekknolagi committed:
Allow 7-byte small strings

tekknolagi committed:
Pattern match on strings

tekknolagi committed:
Bring back some debug collects

tekknolagi committed:
Only emit wrappers for runtime functions that could be used

tekknolagi committed:
Make case labels nicer to read

tekknolagi committed:
Make a compiler web REPL

tekknolagi committed:
Add compiler.py to Docker container for web REPL

tekknolagi commented:
TODO... test it...

tekknolagi committed:
clang-format everything

tekknolagi committed:
clang-tidy everything

tekknolagi committed:
Add note about GC core

gregorybchris commented on README.md:
````diff
@@ -53,6 +53,30 @@ docker run -i -t ghcr.io/tekknolagi/scrapscript:trunk apply "1 + 2"
 docker run -i -t ghcr.io/tekknolagi/scrapscript:trunk repl
 ```
 
+### The experimental compiler:
````
This is so awesome!! What do you think about adding some prelim benchmark results? The first thing I'm wondering is whether it's worth using over the interpreter.

Would understand if you want to hold off until you can benchmark something big or standard.

tekknolagi replied:
Don't have support for variants yet but would like to add them. Shouldn't be too hard. That unlocks a lot of programs

tekknolagi replied:
Then we can benchmark hull.scrap or similar

tekknolagi committed:
Support variants in the compiler

Right now we have an entire 64 bits for the tag and an indirection but
later we may want to use the existing heap object tag's high bits.

tekknolagi committed:
Support printing variants

tekknolagi committed:
Support hole

tekknolagi committed:
Remove unused function

tekknolagi commented:
rebased on top of variants-in-trunk

tekknolagi committed:
Sketch out end-to-end compiler tests

tekknolagi committed:
Partially revert 620d732c6c9e44e5cd4871254c8623df3b57196f

Keep the web REPL interpreter only for now

tekknolagi committed:
Add more tests

tekknolagi committed:
Run compiler tests in CI

tekknolagi committed:
Make sure type annotations work on older Pythons

tekknolagi committed:
Support CC with arguments embedded

tekknolagi committed:
Split the sysconfig var too
